pkg/conference/participant/tracker.go

```diff
@@ -0,0 +1,231 @@
+package participant
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/matrix-org/waterfall/pkg/peer"
+	"github.com/pion/rtp"
+	"github.com/thoas/go-funk"
+)
+
+type (
+	Subscribers      map[ID]*peer.Subscription
+	TrackSubscribers map[TrackID]Subscribers
+)
+
+// Tracks participants and their corresponding tracks.
+// These are grouped together as the field in this structure must be kept synchronized.
+type Tracker struct {
+	participants    map[ID]*Participant
+	subscribers     TrackSubscribers
+	publishedTracks map[TrackID]PublishedTrack
+}
+
+func NewParticipantTracker() *Tracker {
+	return &Tracker{
+		participants:    make(map[ID]*Participant),
+		subscribers:     make(TrackSubscribers),
+		publishedTracks: make(map[TrackID]PublishedTrack),
+	}
+}
+
+// Adds a new participant in the list.
+func (t *Tracker) AddParticipant(participant *Participant) {
+	t.participants[participant.ID] = participant
+}
+
+// Gets an existing participant if any.
+func (t *Tracker) GetParticipant(participantID ID) *Participant {
+	return t.participants[participantID]
+}
+
+func (t *Tracker) HasParticipants() bool {
+	return len(t.participants) != 0
+}
+
+// Iterates over participants and calls a closure on each of the participants.
+func (t *Tracker) ForEachParticipant(fn func(ID, *Participant)) {
+	for id, participant := range t.participants {
+		fn(id, participant)
+	}
+}
+
+// Removes the participant from the conference closing all its tracks. Returns a set of **streams** that are to be
+// removed. The return type is odd since Go does not natively support sets, so we emulate it with a map.
+func (t *Tracker) RemoveParticipant(participantID ID) map[string]bool {
+	participant := t.GetParticipant(participantID)
+	if participant == nil {
+		return make(map[string]bool)
+	}
+
+	// Terminate the participant and remove it from the list.
+	participant.Peer.Terminate()
+	close(participant.HeartbeatPong)
+	delete(t.participants, participantID)
+
+	// Remove the participant's tracks from all participants who might have subscribed to them.
+	streamIdentifiers := make(map[string]bool)
+	for trackID, track := range t.publishedTracks {
+		if track.Owner == participantID {
+			// Odd way to add to a set in Go.
+			streamIdentifiers[track.Info.StreamID] = true
+			t.RemoveTrack(trackID)
+		}
+	}
+
+	return streamIdentifiers
+}
+
+// Adds a new track to the list of published tracks, i.e. by calling it we inform the tracker that there is new track
+// that has been published and that we must take into account from now on.
+func (t *Tracker) AddTrack(
```

Comment:
```suggestion
func (t *Tracker) PublishTrack(
```

Reply:
But we don't publish the track here: the way has been posted at this point already, we add it to the tracker, so we start tracking its state and keep it in sync with the participants' state. Another reason why I chose `AddTrack()` here is to be consistent and symmetric with other functions: we have `add participant`, `remove participant`, `for each participant` and we also have `add track`, `remove track` and `for each track`.

Otherwise having only `add track` called `publish track` might be confusing (i.e. why `publish track`, but `remove track` etc).

Reply:
Right, what about `AddPublishedTrack()`? I wanted there to be a clear distinction between subscribing and adding a published track

Reply:
Ah, but there is no function to mix it with subscribing, i.e. for subscribing there is a `Subscribe` function in a tracker, so adding tracks (`AddTrack()`) is only for published tracks (which we also explain in a comment).

I renamed it to `PublishedTrack` for each function, i.e. `AddTrack` is now `AddPublishedTrack`, and `UpdateTrackMetadata` is now `UpdatePublishedTrackMetadata`. Let me know if it's better (to me it seems a bit too lengthy now, but I don't have a strong opinion on that 🙂).

Reply:
> I renamed it to `PublishedTrack` for each function, i.e. `AddTrack` is now `AddPublishedTrack`, and `UpdateTrackMetadata` is now `UpdatePublishedTrackMetadata`. Let me know if it's better (to me it seems a bit too lengthy now, but I don't have a strong opinion on that ).

Thank you! It's way more readable for me now